[iOS] Support dynamically inserted image analysis context menu items when long pressing image links

whsieh opened the pull request:
#### 24deb829fd303882281b3ef52fa266af13a5e701
<pre>
[iOS] Support dynamically inserted image analysis context menu items when long pressing image links
<a href="https://bugs.webkit.org/show_bug.cgi?id=242577">https://bugs.webkit.org/show_bug.cgi?id=242577</a>
rdar://96800769

Reviewed by Tim Horton.

After 252308@main, when long pressing images inside links in Safari, the contextual image analysis
actions (&quot;Show Text&quot;, &quot;Look Up&quot;, &quot;Copy Subject&quot; and the machine-readable-code submenu) don&apos;t appear
if image analysis takes longer than the context menu appearance animation. This is because we
currently prevent appending these image analysis items in the case where the UI delegate has
supplied its own menu configuration (i.e. the `_canUpdateVisibleContextMenuWithImageAnalysisActions`
flag is set). However, Safari uses this delegate method for image links, and we need to presesrve
the ability for the user to perform these image analysis actions on these links.

To mitigate this, we refactor how this visible menu update codepath works, such that:

1.  If we&apos;re still waiting for image analysis to complete when the context menu shows up, we now
    insert a hidden `UIAction` that represents the list of image analysis items that may potentially
    be inserted once image analysis finishes.

2.  In `-_completeImageAnalysisRequestForContextMenu:`, if this hidden placeholder item is still
    present (in other words, the UI delegate hasn&apos;t explicitly removed it when determining which
    items to present), then swap out this hidden placeholder with the final list of actions.

* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForImageSheet:]):
(-[WKActionSheetAssistant suggestedActionsForContextMenuWithPositionInformation:]):

Make this return an NSMutableArray, so that the content view can append an extra placeholder item
to represent the pending image analysis actions. See below for more details.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:

Replace `_canUpdateVisibleContextMenuWithImageAnalysisActions` with another flag
(`_waitingForDynamicImageAnalysisContextMenuActions`) that indicates whether or not we&apos;re waiting for
image analysis to complete, in order to determine whether the image analysis context menu items
should appear in the long press context menu.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _setUpImageAnalysis]):
(-[WKContentView _tearDownImageAnalysis]):
(-[WKContentView imageAnalysisGestureDidBegin:]):
(-[WKContentView _completeImageAnalysisRequestForContextMenu:requestIdentifier:hasTextResults:]):

Refactor this to search for and replace the placeholder image analysis action with a list of image
analysis items, after analysis has completed. Also, make a minor adjustment to only insert the
&quot;Copy Subject&quot; item if the &quot;Copy&quot; item is also present. Currently, Mail explicitly avoids adding the
&quot;Copy&quot; menu item when long pressing images in managed emails; this tweak allows us to automatically
match that behavior for &quot;Copy Subject&quot;.

(-[WKContentView placeholderForDynamicallyInsertedImageAnalysisActions]):

Add a helper method to create a hidden `UIAction` to append to the recommended list of context menu
actions when long pressing images (by setting the `UIMenuElementAttributesHidden` item attribute).

(-[WKContentView continueContextMenuInteraction:]):
(-[WKContentView _contextMenuInteraction:overrideSuggestedActionsForConfiguration:]):

Canonical link: <a href="https://commits.webkit.org/252356@main">https://commits.webkit.org/252356@main</a>
</pre>

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10973,19 +10973,37 @@ - (void)_completeImageAnalysisRequestForContextMenu:(CGImageRef)image requestIde
             };
 
-            if ([strongSelf copySubjectResultForImageContextMenu])
+            if (foundCopyItem && [strongSelf copySubjectResultForImageContextMenu])
```
Is the idea here that we only want to show "Copy Subject" if we have some sort of general "Copy"?

whsieh replied:
Indeed. I explained the rationale in the commit message, but to summarize — I think it's a sensible default behavior for WebKit clients, and in particular it prevents the Copy Subject action from showing up in (enterprise) managed emails in Mail on iOS.

whsieh committed:
[iOS] Support dynamically inserted image analysis context menu items when long pressing image links
https://bugs.webkit.org/show_bug.cgi?id=242577
rdar://96800769

Reviewed by Tim Horton.

After 252308@main, when long pressing images inside links in Safari, the contextual image analysis
actions ("Show Text", "Look Up", "Copy Subject" and the machine-readable-code submenu) don't appear
if image analysis takes longer than the context menu appearance animation. This is because we
currently prevent appending these image analysis items in the case where the UI delegate has
supplied its own menu configuration (i.e. the `_canUpdateVisibleContextMenuWithImageAnalysisActions`
flag is set). However, Safari uses this delegate method for image links, and we need to presesrve
the ability for the user to perform these image analysis actions on these links.

To mitigate this, we refactor how this visible menu update codepath works, such that:

1.  If we're still waiting for image analysis to complete when the context menu shows up, we now
    insert a hidden `UIAction` that represents the list of image analysis items that may potentially
    be inserted once image analysis finishes.

2.  In `-_completeImageAnalysisRequestForContextMenu:`, if this hidden placeholder item is still
    present (in other words, the UI delegate hasn't explicitly removed it when determining which
    items to present), then swap out this hidden placeholder with the final list of actions.

* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForImageSheet:]):
(-[WKActionSheetAssistant suggestedActionsForContextMenuWithPositionInformation:]):

Make this return an NSMutableArray, so that the content view can append an extra placeholder item
to represent the pending image analysis actions. See below for more details.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:

Replace `_canUpdateVisibleContextMenuWithImageAnalysisActions` with another flag
(`_waitingForDynamicImageAnalysisContextMenuActions`) that indicates whether or not we're waiting for
image analysis to complete, in order to determine whether the image analysis context menu items
should appear in the long press context menu.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _setUpImageAnalysis]):
(-[WKContentView _tearDownImageAnalysis]):
(-[WKContentView imageAnalysisGestureDidBegin:]):
(-[WKContentView _completeImageAnalysisRequestForContextMenu:requestIdentifier:hasTextResults:]):

Refactor this to search for and replace the placeholder image analysis action with a list of image
analysis items, after analysis has completed. Also, make a minor adjustment to only insert the
"Copy Subject" item if the "Copy" item is also present. Currently, Mail explicitly avoids adding the
"Copy" menu item when long pressing images in managed emails; this tweak allows us to automatically
match that behavior for "Copy Subject".

(-[WKContentView placeholderForDynamicallyInsertedImageAnalysisActions]):

Add a helper method to create a hidden `UIAction` to append to the recommended list of context menu
actions when long pressing images (by setting the `UIMenuElementAttributesHidden` item attribute).

(-[WKContentView continueContextMenuInteraction:]):
(-[WKContentView _contextMenuInteraction:overrideSuggestedActionsForConfiguration:]):

Canonical link: https://commits.webkit.org/252356@main
